
![image.png](attachment:image.png)

In [53]:
install.packages('nycflights13')

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



# Lecture 2: More on data transformations

<div style="border: 1px double black; padding: 10px; margin: 10px">

**Goals for today's lecture:**
* Continue to learn how to manipulate data, including:
    * Vectors in R: what they are and basic operations
    * Adding new variables
    * Grouping and summarizing data
    
This lecture note corresponds to sections 4.3-4.6 of your book.
</div>

In [54]:
library(tidyverse)  # always!
library(nycflights13)  # install if needed

## Review of last lecture
Last lecture we learned how to alter the rows and columns of a dataframe:
- `filter()` to keep certain rows that satisfy logical conditions.
- `arrange()` to sort rows according to certain column values.
- `distinct()` to keep only rows that are distinct on some combination of columns.
- `select()` to drop/rename/rearrange columns

## What's a data frame?

Our main goal in R is to work with data, and one of the most fundamental objects in R is the *data frame*. Think of a data frame as a container for a bunch of *vectors* of data:

![dataframe](https://garrettgman.github.io/images/tidy-2.png)

## What's a vector?

- In programming speak: a *vector* is a list of values. 
- In statistical speak: a vector of observations (aka data).

Let's create a vector and work with it:

## Poll
How old are you?
<ol style="list-style-type: upper-alpha;">
    <li>19 or younger</li>
    <li>20</li>
    <li>21</li>
    <li>22 or older</li>
    <li>I forget</li>
</ol>
(This question will be graded.)<br />
(↑ This is a joke.)

The function for creating a vector in R is called simply, `c()`.

In [56]:
# Create a vector of ages
ages = c(20, 21, 22, 22, 24, 25, 21, 20, 19, 18)
ages 

[1] 20 21 22 22 24 25 21 20 19 18

## Functions that operate on vectors

Many summary functions are available:

* Center: `mean(), median()`
* Spread: `sd(), IQR(), mad()`
* Range: `min(), max(), quantile()`
* Position: `first(), last(), nth()`
* Count: `n(), n_distinct()`
* Logical: `any(), all()`

In [57]:
log1 = c(T, T, T)
all(log1)

[1] TRUE

In [61]:
# examples of functions we can use on the ages vector
print(sd(ages))
print(mean(ages))

[1] 2.149935
[1] 21.2


## Column operation #2: `mutate()`

`mutate()` creates new columns in a dataframe that are calculated from the existing columns.

For example, let's define the **gain** of a flight to be the different between the departure delay and the arrival delay:

$$\text{gain} = \text{dep. delay} - \text{arr. delay}$$

So, the gain is positive if the flight made up time in the air, resulting in a less-delayed arrival.

To add a column called `gain` to flights, we called `mutate()` as follows:

In [60]:
# add a gain column to flights
# check if gain column is not there
select(flights, gain)

ERROR: ignored

In [62]:
new_df = mutate(flights, gain = dep_delay - arr_delay)

In [63]:
head(new_df, 1)

year month day dep_time sched_dep_time dep_delay arr_time sched_arr_time
1 2013 1     1   517      515            2         830      819           
  arr_delay carrier flight tailnum origin dest air_time distance hour minute
1 11        UA      1545   N14228  EWR    IAH  227      1400     5    15    
  time_hour           gain
1 2013-01-01 05:00:00 -9

The returned data set has a new column called `gain` added to the very end. If you want to bring it to the front, you could use `select()` like we learned last lecture:

In [64]:
# use select to move gain, dep_delay, arr_delay to the set of columns

print(select(new_df, gain, dep_delay, arr_delay, everything()))


# A tibble: 336,776 × 20
    gain dep_delay arr_delay  year month   day dep_time sched_…¹ arr_t…² sched…³
   <dbl>     <dbl>     <dbl> <int> <int> <int>    <int>    <int>   <int>   <int>
 1    -9         2        11  2013     1     1      517      515     830     819
 2   -16         4        20  2013     1     1      533      529     850     830
 3   -31         2        33  2013     1     1      542      540     923     850
 4    17        -1       -18  2013     1     1      544      545    1004    1022
 5    19        -6       -25  2013     1     1      554      600     812     837
 6   -16        -4        12  2013     1     1      554      558     740     728
 7   -24        -5        19  2013     1     1      555      600     913     854
 8    11        -3       -14  2013     1     1      557      600     709     723
 9     5        -3        -8  2013     1     1      557      600     838     846
10   -10        -2         8  2013     1     1      558      600     753     745
# …

## Quiz
What was the most amount of time gained by any flight?
<ol style="list-style-type: upper-alpha;">
    <li>2 hours</li>
    <li>109 minutes</li>
    <li>37 minutes</li>
    <li>37 seconds</li>
    <li>2 days</li>
</ol>

In [66]:
# Add arrange to see the gain in desc order and head function shows us the first 6 records by default
head(
  arrange(
    select(new_df, gain, dep_delay, arr_delay, everything()), 
    desc(gain)
), 5)

gain dep_delay arr_delay year month day dep_time sched_dep_time arr_time
1 109  235       126       2013 6     13  1907     1512           2134    
2  87   60       -27       2013 2     26  1000      900           1513    
3  80  206       126       2013 2     23  1226      900           1746    
4  79   17       -62       2013 5     13  1917     1900           2149    
5  76   24       -52       2013 2     27   924      900           1448    
  sched_arr_time carrier flight tailnum origin dest air_time distance hour
1 1928           EV      4377   N19554  EWR    JAX  126       820     15  
2 1540           HA        51   N382HA  JFK    HNL  584      4983      9  
3 1540           HA        51   N389HA  JFK    HNL  599      4983      9  
4 2251           DL      1465   N721TW  JFK    SFO  313      2586     19  
5 1540           HA        51   N389HA  JFK    HNL  589      4983      9  
  minute time_hour          
1 12     2013-06-13 15:00:00
2  0     2013-02-26 09:00:00
3  0     2013-02-23 09:00:00
4  0     2013-05-13 19:00:00
5  0     2013-02-27 09:00:00

## Filtering extreme values
In the previous question we needed to find rows that had a large value of a certain column (`gain`). This occurs frequently, so the designers of tidyverse wrote a special function:

    top_n(<DATA FRAME>, n, <COLUMN>, ...)
    

In [67]:
# use top_n to find the flights with the highest gain
top_n(new_df, 1, gain)

year month day dep_time sched_dep_time dep_delay arr_time sched_arr_time
1 2013 6     13  1907     1512           235       2134     1928          
  arr_delay carrier flight tailnum origin dest air_time distance hour minute
1 126       EV      4377   N19554  EWR    JAX  126      820      15   12    
  time_hour           gain
1 2013-06-13 15:00:00 109

## Grouping data
Very frequently our data have natural groupings. For example, in flights, we might be interested in studying differences in flights depending on the month of departure. We use the `group_by()` function to tell R how to group data.

For example, `mtcars` is a dataset of cars and the gas mileage they get:

In [68]:
head(mtcars, 5) %>% print

                   mpg cyl disp  hp drat    wt  qsec vs am gear carb
Mazda RX4         21.0   6  160 110 3.90 2.620 16.46  0  1    4    4
Mazda RX4 Wag     21.0   6  160 110 3.90 2.875 17.02  0  1    4    4
Datsun 710        22.8   4  108  93 3.85 2.320 18.61  1  1    4    1
Hornet 4 Drive    21.4   6  258 110 3.08 3.215 19.44  1  0    3    1
Hornet Sportabout 18.7   8  360 175 3.15 3.440 17.02  0  0    3    2


In [69]:
help(mtcars)

Let's try grouping `mtcars` by `cyl` (the number of engine cylinders):

In [70]:
gp = group_by(mtcars, cyl)
print(gp)

# A tibble: 32 × 11
# Groups:   cyl [3]
     mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb
   <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
 1  21       6  160    110  3.9   2.62  16.5     0     1     4     4
 2  21       6  160    110  3.9   2.88  17.0     0     1     4     4
 3  22.8     4  108     93  3.85  2.32  18.6     1     1     4     1
 4  21.4     6  258    110  3.08  3.22  19.4     1     0     3     1
 5  18.7     8  360    175  3.15  3.44  17.0     0     0     3     2
 6  18.1     6  225    105  2.76  3.46  20.2     1     0     3     1
 7  14.3     8  360    245  3.21  3.57  15.8     0     0     3     4
 8  24.4     4  147.    62  3.69  3.19  20       1     0     4     2
 9  22.8     4  141.    95  3.92  3.15  22.9     1     0     4     2
10  19.2     6  168.   123  3.92  3.44  18.3     1     0     4     4
# … with 22 more rows


This has not changed the data in any way. But now watch what happens when we use `mutate()` on the grouped data frame:

In [71]:
# mean mpg for grouped data
print(mutate(gp, mean(mpg)))

# A tibble: 32 × 12
# Groups:   cyl [3]
     mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb `mean(mpg)`
   <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>       <dbl>
 1  21       6  160    110  3.9   2.62  16.5     0     1     4     4        19.7
 2  21       6  160    110  3.9   2.88  17.0     0     1     4     4        19.7
 3  22.8     4  108     93  3.85  2.32  18.6     1     1     4     1        26.7
 4  21.4     6  258    110  3.08  3.22  19.4     1     0     3     1        19.7
 5  18.7     8  360    175  3.15  3.44  17.0     0     0     3     2        15.1
 6  18.1     6  225    105  2.76  3.46  20.2     1     0     3     1        19.7
 7  14.3     8  360    245  3.21  3.57  15.8     0     0     3     4        15.1
 8  24.4     4  147.    62  3.69  3.19  20       1     0     4     2        26.7
 9  22.8     4  141.    95  3.92  3.15  22.9     1     0     4     2        26.7
10  19.2     6  168.   123  3.92  3.44  18.3     1     0     4     4 

In [72]:
distinct(mtcars, cyl)

cyl
Mazda RX4         6  
Datsun 710        4  
Hornet Sportabout 8

Notice that the mean is now constant within different groups. It's easier to see if we first sort the table by `cyl`:

In [73]:
# sort mtcars by cyl, then group and mutate
arrange(mutate(gp, mean(mpg)), cyl) %>% print

# A tibble: 32 × 12
# Groups:   cyl [3]
     mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb `mean(mpg)`
   <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>       <dbl>
 1  22.8     4 108      93  3.85  2.32  18.6     1     1     4     1        26.7
 2  24.4     4 147.     62  3.69  3.19  20       1     0     4     2        26.7
 3  22.8     4 141.     95  3.92  3.15  22.9     1     0     4     2        26.7
 4  32.4     4  78.7    66  4.08  2.2   19.5     1     1     4     1        26.7
 5  30.4     4  75.7    52  4.93  1.62  18.5     1     1     4     2        26.7
 6  33.9     4  71.1    65  4.22  1.84  19.9     1     1     4     1        26.7
 7  21.5     4 120.     97  3.7   2.46  20.0     1     0     3     1        26.7
 8  27.3     4  79      66  4.08  1.94  18.9     1     1     4     1        26.7
 9  26       4 120.     91  4.43  2.14  16.7     0     1     5     2        26.7
10  30.4     4  95.1   113  3.77  1.51  16.9     1     1     5     2 

## Pipes
I have already used the pipe symbol, we will make extensive use of the pipe operator `%>%` going forward. Consider the previous exercise:

In [74]:
# mutate flights to add gain column, select gain, dep_delay and arr_delay and finally sort it by gain and then print
print(arrange(
  select(mutate(flights, gain = dep_delay - arr_delay), 
  gain, dep_delay, arr_delay, everything()
  ), desc(gain)
))

# A tibble: 336,776 × 20
    gain dep_delay arr_delay  year month   day dep_time sched_…¹ arr_t…² sched…³
   <dbl>     <dbl>     <dbl> <int> <int> <int>    <int>    <int>   <int>   <int>
 1   109       235       126  2013     6    13     1907     1512    2134    1928
 2    87        60       -27  2013     2    26     1000      900    1513    1540
 3    80       206       126  2013     2    23     1226      900    1746    1540
 4    79        17       -62  2013     5    13     1917     1900    2149    2251
 5    76        24       -52  2013     2    27      924      900    1448    1540
 6    74        48       -26  2013     7    14     1917     1829    2109    2135
 7    74        34       -40  2013     7    17     2004     1930    2224    2304
 8    73        31       -42  2013    12    27     1719     1648    1956    2038
 9    73        -2       -75  2013     5     2     1947     1949    2209    2324
10    72         9       -63  2013    11    13     2024     2015    2251    2354
# …

This is not very nice. To figure out what the command is doing you have to work from the inside out, which is not the order in which we are accustomed to reading. A slight improvement might be:

In [75]:
# mutate flights to add gain column, select gain, dep_delay and arr_delay and finally sort it by gain and then print
new_df1 = mutate(flights, gain = dep_delay - arr_delay)
new_df2 = select(new_df1, gain, dep_delay, arr_delay, everything())
new_df3 = arrange(new_df2, desc(gain))
print(new_df3)

# A tibble: 336,776 × 20
    gain dep_delay arr_delay  year month   day dep_time sched_…¹ arr_t…² sched…³
   <dbl>     <dbl>     <dbl> <int> <int> <int>    <int>    <int>   <int>   <int>
 1   109       235       126  2013     6    13     1907     1512    2134    1928
 2    87        60       -27  2013     2    26     1000      900    1513    1540
 3    80       206       126  2013     2    23     1226      900    1746    1540
 4    79        17       -62  2013     5    13     1917     1900    2149    2251
 5    76        24       -52  2013     2    27      924      900    1448    1540
 6    74        48       -26  2013     7    14     1917     1829    2109    2135
 7    74        34       -40  2013     7    17     2004     1930    2224    2304
 8    73        31       -42  2013    12    27     1719     1648    1956    2038
 9    73        -2       -75  2013     5     2     1947     1949    2209    2324
10    72         9       -63  2013    11    13     2024     2015    2251    2354
# …

This is better, but now you've created a bunch of useless temporary variables, and it requires a lot of typing. 
Instead, we are going to use a new operator `%>%` (prounouced "pipe"):

In [77]:
# # mutate flights to add gain column, select gain, dep_delay and arr_delay and finally sort it by gain and then print using pipes
mutate(flights, gain = dep_delay - arr_delay) %>% 
  select(gain, dep_delay, arr_delay, everything()) %>% 
    arrange(desc(gain)) %>%
      print
      

# A tibble: 336,776 × 20
    gain dep_delay arr_delay  year month   day dep_time sched_…¹ arr_t…² sched…³
   <dbl>     <dbl>     <dbl> <int> <int> <int>    <int>    <int>   <int>   <int>
 1   109       235       126  2013     6    13     1907     1512    2134    1928
 2    87        60       -27  2013     2    26     1000      900    1513    1540
 3    80       206       126  2013     2    23     1226      900    1746    1540
 4    79        17       -62  2013     5    13     1917     1900    2149    2251
 5    76        24       -52  2013     2    27      924      900    1448    1540
 6    74        48       -26  2013     7    14     1917     1829    2109    2135
 7    74        34       -40  2013     7    17     2004     1930    2224    2304
 8    73        31       -42  2013    12    27     1719     1648    1956    2038
 9    73        -2       -75  2013     5     2     1947     1949    2209    2324
10    72         9       -63  2013    11    13     2024     2015    2251    2354
# …

This is much better. We can read the command from left to right and know exactly what is going on.

## Column operation #3: `summarize()`ing data

`summarize()` can be used to summarize entire data frames by collapsing them into single number summaries. The syntax is:

    summarize(<grouped data frame>, 
              <new variable> = <formula for new variable>,
              <other new variable> = <other formula>)

The most basic use of summarize is to compute statistics over the whole data set:

In [78]:
# summarize flights by mean of departure delay
# note, if you don't add na.rm=T; which removes all NA values before computing the mean 

summarise(flights, mean_dep_delay = mean(dep_delay, na.rm = T)) 

mean_dep_delay
1 12.63907

`summarize()` applies a summary function to each group of data. Remember that it always returns **one row per group**. In the above example, there was only one group (the whole data set), so the resulting data frame had only one row.

### Grouping observations
`summarize()` is most useful when combined with `group_by()` to group observations before calculating the summary statistic. Let's summarize flights by the mean departure delay in each month.

In [79]:
# summarize average departure delay by month.
group_by(flights, month) %>%
  summarise(mean_dep_delay = mean(dep_delay, na.rm = T)) %>%
  print

# A tibble: 12 × 2
   month mean_dep_delay
   <int>          <dbl>
 1     1          10.0 
 2     2          10.8 
 3     3          13.2 
 4     4          13.9 
 5     5          13.0 
 6     6          20.8 
 7     7          21.7 
 8     8          12.6 
 9     9           6.72
10    10           6.24
11    11           5.44
12    12          16.6 


In [80]:
# summarize average mpg of mtcars.
# In the class this was shown
group_by(mtcars, cyl) %>% 
  summarise(mean_mpg = mean(mpg)) %>% 
    print

# A tibble: 3 × 2
    cyl mean_mpg
  <dbl>    <dbl>
1     4     26.7
2     6     19.7
3     8     15.1


### Example: counting the number of rows
The `n()` function calculates the number of rows in each group:

In [81]:

group_by(mtcars, cyl) %>% 
  summarise(mean_mpg = mean(mpg), n = n()) %>% 
    print

# A tibble: 3 × 3
    cyl mean_mpg     n
  <dbl>    <dbl> <int>
1     4     26.7    11
2     6     19.7     7
3     8     15.1    14


In [82]:
# count the number of rows in flights for each month
group_by(flights, month) %>%
  summarise(mean_dep_delay = mean(dep_delay, na.rm = T), n = n()) %>%
  print

# A tibble: 12 × 3
   month mean_dep_delay     n
   <int>          <dbl> <int>
 1     1          10.0  27004
 2     2          10.8  24951
 3     3          13.2  28834
 4     4          13.9  28330
 5     5          13.0  28796
 6     6          20.8  28243
 7     7          21.7  29425
 8     8          12.6  29327
 9     9           6.72 27574
10    10           6.24 28889
11    11           5.44 27268
12    12          16.6  28135


### A shortcut
`summarize(n = n())` occurs so often that there is a shortcut for it:

In [83]:
# use count() instead of group_by or summarize
count(mtcars, cyl)

cyl n 
1 4   11
2 6    7
3 8   14

Let's think about how to answer the following question using `summarize`:

## What days of the year / at what airport are the busiest for flying?

To figure this out, I like to think about/visualize the table we would want to have in order to easily answer this question. Ideally, it would look something like this:

    # A tibble: 1,095 x 4
       month   day airport       n_departures
       <int> <int> <chr>                <int>
     1     1     1 EWR                    305
     2     1     1 JFK                    297
     3     1     1 LGA                    240
     4     1     2 EWR                    350
     5     1     2 JFK                    321
     6     1     2 LGA                    272
     7     1     3 EWR                    336
     8     1     3 JFK                    318
     9     1     3 LGA                    260
    10     1     4 EWR                    339
    # … with 1,085 more rows

Then, to get the answer, I could sort the table to find the row that had the largest `n_departures`.

How do I reach the table shown above? There is one row per ... what? (This tells me how to group the data.)

In [84]:
# summarize flights to get number of departures by day and by airport.
group_by(flights, month, day, origin) %>% 
  summarise(n = n()) %>% 
    print

`summarise()` has grouped output by 'month', 'day'. You can override using the
`.groups` argument.


# A tibble: 1,095 × 4
# Groups:   month, day [365]
   month   day origin     n
   <int> <int> <chr>  <int>
 1     1     1 EWR      305
 2     1     1 JFK      297
 3     1     1 LGA      240
 4     1     2 EWR      350
 5     1     2 JFK      321
 6     1     2 LGA      272
 7     1     3 EWR      336
 8     1     3 JFK      318
 9     1     3 LGA      260
10     1     4 EWR      339
# … with 1,085 more rows


Here is another question we can answer:

## Who is the greatest (baseball) batter of all time?
The `Lahman` dataset contains information on baseball players.

In [85]:
install.packages("Lahman")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [86]:
# install.packages("Lahman")
library(Lahman)
as_tibble(Batting) %>% print
# what do all these columns mean?

# A tibble: 110,495 × 22
   playerID  yearID stint teamID lgID      G    AB     R     H   X2B   X3B    HR
   <chr>      <int> <int> <fct>  <fct> <int> <int> <int> <int> <int> <int> <int>
 1 abercda01   1871     1 TRO    NA        1     4     0     0     0     0     0
 2 addybo01    1871     1 RC1    NA       25   118    30    32     6     0     0
 3 allisar01   1871     1 CL1    NA       29   137    28    40     4     5     0
 4 allisdo01   1871     1 WS3    NA       27   133    28    44    10     2     2
 5 ansonca01   1871     1 RC1    NA       25   120    29    39    11     3     0
 6 armstbo01   1871     1 FW1    NA       12    49     9    11     2     1     0
 7 barkeal01   1871     1 RC1    NA        1     4     0     1     0     0     0
 8 barnero01   1871     1 BS1    NA       31   157    66    63    10     9     0
 9 barrebi01   1871     1 FW1    NA        1     5     1     1     1     0     0
10 barrofr01   1871     1 BS1    NA       18    86    13    13     2     1     0
# …

The second player is `addybo01`. We can get information about this player by typing:

In [87]:
Lahman::playerInfo('addybo01')

playerID nameFirst nameLast
111 addybo01 Bob       Addy

Bob Addy was active in the years 1871-1877. During that time he had $118+51+152+213+310+142+245=1231$ at-bats, and $32+16+54+51+80+40+68=341$ hits. Therefore his career batting average was $341/1241=0.277$.

In [89]:
filter(Batting, playerID == 'addybo01') %>% print

  playerID yearID stint teamID lgID  G  AB  R  H X2B X3B HR RBI SB CS BB SO IBB
1 addybo01   1871     1    RC1   NA 25 118 30 32   6   0  0  13  8  1  4  0  NA
2 addybo01   1873     1    PH2   NA 10  51 12 16   1   0  0  10  1  1  2  0  NA
3 addybo01   1873     2    BS1   NA 31 152 37 54   6   3  1  32  6  5  2  1  NA
4 addybo01   1874     1    HR1   NA 50 213 25 51   9   2  0  22  4  2  1  1  NA
5 addybo01   1875     1    PH2   NA 69 310 60 80   8   4  0  43 16  8  0  2  NA
6 addybo01   1876     1    CHN   NL 32 142 36 40   4   1  0  16 NA NA  5  0  NA
7 addybo01   1877     1    CN1   NL 57 245 27 68   2   3  0  31 NA NA  6  5  NA
  HBP SH SF GIDP
1  NA NA NA    0
2  NA NA NA    0
3  NA NA NA    0
4  NA NA NA    0
5  NA NA NA    0
6  NA NA NA   NA
7  NA NA NA   NA


Let's use `group_by()` and `summarize()` to calculate the "career" batting average for every player in the dataset. That is, I want a table that looks like:

    # A tibble: 20166 × 2
      playerID batting_avg
      <chr>          <dbl>
    1 addybo01       0.277
    .    .             .
    .    .             .
    .    .             .

In [90]:
# calculate the batting average for each player in the  data set
group_by(Batting, playerID) %>% 
  summarise('batting avg' = sum(H)/sum(AB)) %>% 
    top_n(10) %>%
      print

Selecting by batting avg


# A tibble: 95 × 2
   playerID  `batting avg`
   <chr>             <dbl>
 1 abramge01             1
 2 alberan01             1
 3 banisje01             1
 4 bartocl01             1
 5 bassdo01              1
 6 birasst01             1
 7 bruneju01             1
 8 burnscb01             1
 9 cammaer01             1
10 campsh01              1
# … with 85 more rows


What has happened? Let's look at the first player in the sorted table:

In [91]:
filter(Batting, playerID == 'abramge01')

playerID  yearID stint teamID lgID G AB R H X2B ⋯ RBI SB CS BB SO IBB HBP SH
1 abramge01 1923   1     CIN    NL   3 1  0 1 0   ⋯ 0   0  0  0  0  NA  0   0 
  SF GIDP
1 NA NA

### Always include counts
It is a good idea to include counts of each group when you do a summary. Some groups may have very low numbers of observations, resulting in high variance for the summary statistics. 

What happens if we restrict our batting average calculation to players that had at least 100 at-bats?

## Quiz
Among players who had at least 100 at bats, who had the highest career batting average?
<ol style="list-style-type: upper-alpha;">
    <li>Ty Cobb</li>
    <li>Babe Ruth</li>
    <li>Prof. Terhorst</li>
    <li>Ted Williams</li>
    <li>Hank Williams</li>
</ol>

In [92]:
# highest batting average among players that had 100 or more at bats
filter(Batting, AB >= 100) %>%
  group_by(playerID) %>% 
  summarise(batting_avg = sum(H)/sum(AB)) %>% 
    arrange(desc(batting_avg)) %>%
      print

# A tibble: 7,158 × 2
   playerID  batting_avg
   <chr>           <dbl>
 1 hazlebo01       0.403
 2 daviscu01       0.381
 3 fishesh01       0.374
 4 woltery01       0.370
 5 cobbty01        0.366
 6 barnero01       0.363
 7 hornsro01       0.362
 8 meyerle01       0.358
 9 jacksjo01       0.357
10 harveza01       0.353
# … with 7,148 more rows


## Quiz
Among players who had at least 100 at bats in a season, what was the highest batting average in a single season?
<ol style="list-style-type: upper-alpha;">
    <li>Ted Williams</li>
    <li>Steven Colbert</li>
    <li>Chonky Squirrel</li>
    <li>Levi Meyerle</li>
    <li>Tom Riddle</li>
</ol>

In [ ]:
# highest seasonal batting average
# I will let you folks figure this out

## The Steroid Era of Baseball

> [Baseball] remained relatively the same until the 90s when steroid use became rampant. Famous sluggers like Barry Bonds, Mark McGwire, and Sammy Sosa rose to fame during this era. They were beloved at the time until we later found out that they were cheating.

https://www.wagerbop.com/how-home-runs-and-batting-averages-have-changed-over-the-last-30-years/

![barry bounds](https://cdn.ebaumsworld.com/mediaFiles/picture/2605038/87087115.jpg)

## Can we see the steroid era reflected in the data?

In [ ]:
# summarize the dataset in order to investigate steroid era in batters
# this one too - exercise for you folks